pulsar-client/src/main/java/org/apache/pulsar/client/api/PulsarClient.java

```diff
@@ -73,7 +73,9 @@ public static PulsarClient create(String serviceUrl, ClientConfiguration conf) t
      * @throws PulsarClientException.AuthorizationException
      *             if the authorization to publish on topic was denied
      */
-    Producer createProducer(String topic) throws PulsarClientException;
+    Producer<byte[]> createProducer(String topic) throws PulsarClientException;
+
+    <T> Producer<T> createProducer(String topic, Codec<T> codec) throws PulsarClientException;
```

Comment:
I like the syntax!

About `Codec` should we name it actually something like `Schema` to imply this will not be used just for serialize/deserialize but also for schema validation?

Reply:
For example, I was thinking something like : 

```java
class MyClass {
  public int x;
  public int y;
}

Producer<MyClass> producer = client.createProducer(topic,
                                       JSONSchema.of(MyClass.class));

// or 
Producer<MyClass> producer = client.createProducer(topic,
                                 AvroSchema.of(MyClass.class));
```
Another syntax option:

```java
Producer<MyClass> producer = client.createProducer(topic,
                                 Schema.Avro(MyClass.class));
```

The key difference would be to always imply a schema (whether that is enforced or not in server side is a different matter), rather that implying just the serder codec which will be anyway specified by the schema.

Of course, that would make it more difficult to use typed API with a custom codec (eg: no avro/json or something natively supported), but I believe it's a better semantic overall.

Reply:
This seems reasonable. Let me work on this.